src/rgw/rgw_rest_log.cc

```diff
@@ -605,19 +606,24 @@ void RGWOp_DATALog_List::execute() {
       return;
     }
   } 
-  
+
+  dout(10) << __func__ << " shard " << shard << " st " << st << " et " << et << " marker " 
+         << marker << " max_entries "<< max_entries << dendl;
+  left_cnt = max_entries;
+
   do {
     // Note that last_marker is updated to be the marker of the last
     // entry listed
     http_ret = store->data_log->list_entries(shard_id, ut_st, ut_et, 
-					     max_entries, entries, marker,
+					     left_cnt, entries, marker,
 					     &last_marker, &truncated);
     if (http_ret < 0) 
       break;
 
-    if (!max_entries_str.empty()) 
-      max_entries -= entries.size();
-  } while (truncated && (max_entries > 0));
+    assert(max_entries >= entries.size());
+    left_cnt = max_entries - entries.size();
```

Comment:
could you please make a similar change in `RGWOp_MDLog_List`? it appears to handle `last_marker` correctly, but the `if (!max_entries_str.empty())` there should go away too